aten/src/ATen/native/mkldnn/Conv.cpp

```diff
@@ -35,43 +35,84 @@ std::tuple<at::Tensor,at::Tensor,at::Tensor> mkldnn_convolution_backward(
+    AT_ASSERTM(
+        groups > 1,
+        "Only group _mkldnn_conv2d weights could have been reordered to 5d");
+    kernel_size[0] = weight.size(0) * weight.size(1);
+    std::copy_n(
+        weight.sizes().cbegin() + 2, input.dim() - 1, kernel_size.begin() + 1);
+  } else {
+    std::copy_n(
+        weight.sizes().cbegin(), input.dim(), kernel_size.begin());
+  }
 
-// Often written as 2 + max_dim (extra dims for batch size and channels)
-constexpr int max_dim = 3;
+  std::vector<int64_t> output_sizes = conv_output_size(
+    input.sizes(), kernel_size, padding, stride, dilation);
 
-static std::vector<int64_t> conv_output_size(
-    IntArrayRef input_size, IntArrayRef weight_size,
-    IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation, int64_t groups)
-{
-  auto dim = input_size.size();
-  std::vector<int64_t> output_size(dim);
-  output_size[0] = input_size[input_batch_size_dim];
-  output_size[1] = weight_size[weight_output_channels_dim];
-  for (size_t d = 2; d < dim; ++d) {
-    auto kernel = dilation[d - 2] * (weight_size[d] - 1) + 1;
-    output_size[d] = (input_size[d] + (2 * padding[d - 2])
-                        - kernel) / stride[d - 2] + 1;
+  ideep::tensor y;
+  if (bias.defined()) {
+    const ideep::tensor& b = itensor_from_mkldnn(bias);
+    ideep::convolution_forward::compute<AllocForMKLDNN>(
+      x,
+      w,
+      b,
+      {output_sizes.cbegin(), output_sizes.cend()},
+      y,
+      {stride.begin(), stride.end()},
+      {dilation.begin(), dilation.end()},
+      {padding.begin(), padding.end()},
+      {padding.begin(), padding.end()},
+      groups,
+      ideep::descriptor_group::attr_t{},
+      ideep::algorithm::convolution_direct,
+      ideep::prop_kind::forward);
+  } else {
+    ideep::convolution_forward::compute<AllocForMKLDNN>(
+      x,
+      w,
+      {output_sizes.cbegin(), output_sizes.cend()},
+      y,
+      {stride.begin(), stride.end()},
+      {dilation.begin(), dilation.end()},
+      {padding.begin(), padding.end()},
+      {padding.begin(), padding.end()},
+      groups,
+      ideep::descriptor_group::attr_t{},
+      ideep::algorithm::convolution_direct,
+      ideep::prop_kind::forward);
   }
-  return output_size;
+  return new_with_itensor_mkldnn(std::move(y), input.options());
 }
 
 at::Tensor mkldnn_convolution(
     const at::Tensor& input, const at::Tensor& weight, const at::Tensor& bias,
     IntArrayRef padding, IntArrayRef stride, IntArrayRef dilation, int64_t groups)
 {
-  auto output = at::empty(conv_output_size(
-    input.sizes(), weight.sizes(), padding, stride, dilation, groups), input.options());
+  if (input.type_id() == MkldnnCPUTensorId()) {
+    return _mkldnn_conv2d(input, weight, bias, padding, stride, dilation, groups);
+  }
+
+  auto output = at::empty(
+      conv_output_size(
+          input.sizes(), weight.sizes(), padding, stride, dilation),
+      input.options());
 
   auto cpu_engine = CpuEngine::Instance().get_engine();
```

Comment:
we should change the code below using new stuff too (i.e. to_mkldnn -> conv -> to_dense), but it can be a separate PR. Add a TODO?

Reply:
Yes we should.
In fact if you look at the previous commit history I did that, but then somehow this path becomes slower (comparing to the existing direct mkldnn code) :(, I haven't spent time to investigate it yet, but we should definitely do it (in a separate PR).